Implement Presence Channels

SidneyNemzer opened the pull request:
## Related Tickets/PRs

https://goguardian.atlassian.net/browse/GGT-4010

## Summary of Changes

This PR implements presence channels by providing an API for consumers to access channel members and listen for member changes.

## Additional Information

```
type PresenceChannel interface {
  // PresenceChannel inherits Channel functions
  Channel

  // BindMemberAdded returns a channel that receives a Member value when a user
  // joins the channel.
  BindMemberAdded() chan Member

  // BindMemberRemoved returns a channel that receives a user ID when a user
  // leaves the channel.
  BindMemberRemoved() chan string

  // Members returns the member info of each user currently subscribed to the
  // channel.
  Members() map[string]Member

  // Member returns a representation of the channel member with the given ID.
  // If the member isn't in the channel, the returned bool is `false`.
  Member(id string) (bool, Member)

  // Me returns the member for the current user.
  //
  // Possible errors:
  //  - not subscribed - subscription must succeed before calling Me()
  //  - invalid channel data - pusher server provided invalid JSON
  //  - missing member for current user - pusher server violated protocol
  Me() (*Member, error)

  // MemberCount returns the number of users connected to the channel.
  MemberCount() int
}
```

## Checklist

- [x] unit tests
  90% coverage for `presence_channel.go`. The uncovered code is the error branches.

SidneyNemzer committed:
go mod init

SidneyNemzer committed:
implement presence channels

Also modifies all channels so that Unbind()-ing a message channel cleans
up any goroutines that were blocked waiting to send on that channel.

SidneyNemzer committed:
create listen example

SidneyNemzer committed:
remove bitstamp example

bitstamp no longer uses pusher

anatolym commented on presence_channel.go:
```diff
@@ -0,0 +1,249 @@
+	Members() map[string]Member
+	// Member returns a representation of the channel member with the given ID.
+	// If the member isn't in the channel, the returned bool is `false`.
+	Member(id string) (bool, Member)
```
If possible, I'd suggest to use more common signature for this method like
- `Member(id string) (*Member)` where bool = false case is replaced with nil Member, or
- `Member(id string) (Member, bool)` similar to go's `value, ok := someMap[key]`

Also a note. Currently this method returns a shallow copy of a Member which has both a copy and a reference to the original instance.

```go
type Member struct {
	ID   string
	Info json.RawMessage
}
```

When `Member` method returns a copy, it creates new instance of Member, where `ID` is a copy of original string `ID`, but `Info` will be a reference to original `Info`. That's because underlying type of `json.RawMessage` is `[]byte`.

So depending on the use case, we should
a) return a reference to the original member (not creating a shallow copy),
b) return a deep copy of the original member (requires allocating new `json.RawMessage`),
c) return a shallow copy (current implementation) and have a comment about this in the method description.

SidneyNemzer replied:
I like `Member(id string) (*Member)`, this would avoid copying and simplifies the signature. I'll add a note that `Member` shouldn't be modified. I'll also modify the other Member methods to return `*Member` to keep things consistent.

anatolym commented on presence_channel.go:
```diff
@@ -0,0 +1,249 @@
+		default:
+		}
+
+		members := map[string]Member{}
```
nit:
```suggestion
		members := make(map[string]Member, len(subscriptionData.Presence.Hash))
```

anatolym commented on presence_channel.go:
```diff
@@ -0,0 +1,249 @@
+
+	// Maps are passed by reference, so a copy must be made to avoid giving the
+	// caller a reference to the internal map.
+	members := map[string]Member{}
```
nit:
```suggestion
	members := make(map[string]Member, len(c.members))
```

anatolym commented on presence_channel.go:
```diff
@@ -0,0 +1,249 @@
+	defer c.mutex.RUnlock()
+
+	if !c.subscribed {
+		return nil, errors.New("not subscribed")
```
Should we define this error as an exported named variable for easier assertion on the caller side? Similar to `ErrTimedOut` https://github.com/goguardian/pusher-ws-go/blob/f0e8865b9e8db7567d2ae8265fd9d3c7fe9c6803/channel.go#L91

anatolym commented on presence_channel.go:
```diff
@@ -0,0 +1,249 @@
+		fmt.Println(string(c.channelData.ChannelData))
+		// This indicates the pusher server violated the protocol; it did not
+		// provide member info for the current user
+		return nil, errors.New("missing member for current user")
```
ditto: named error

SidneyNemzer committed:
use NewTimer instead of After

This allows the timer to be cleaned up if the subscription succeeds
before the timeout.

anatolym commented on client.go:
```diff
@@ -268,6 +270,22 @@ func (c *Client) Subscribe(channelName string, opts ...SubscribeOption) (Channel
+		return nil, fmt.Errorf("invalid presence channel name, must start with 'presence-': %s", channelName)
+	}
+
+	ch, subscribeErr := c.Subscribe(channelName, opts...)
+	return ch.(*presenceChannel), subscribeErr
```
I have a suspicion that this may panic when `c.Subscribe` returns `nil, <some error>`. I tried replicating [here](https://go.dev/play/p/fPZ2yf4RWLx), but I'm not sure that's correct example.

Can we pls make it explicit like
```suggestion
	ch, subscribeErr := c.Subscribe(channelName, opts...)
	if err != nil {
		return nil, subscribeErr
	}
	
	// <comment about why this will never panic or another explicit type cast and error handling>
	return ch.(*presenceChannel), nil
```

anatolym replied:
Can we add a test case to confirm this?

SidneyNemzer replied:
`ch` won't ever be `nil` here, a channel struct is always created and returned although it may not successfully subscribe. I'll update the doc comment to make that clear.

dkarella commented on presence_channel.go:
```diff
@@ -0,0 +1,249 @@
+		if err != nil {
+			c.client.sendError(err)
+		} else {
+			c.mutex.Lock()
```
Could you make sure that locking here doesn't result in a deadlock in actual usage? 

The reason I suspect this happens is because the caller of `handleEvent`, `Client.listen`, obtains an RLock prior to the call. Accoding to [the docs](https://pkg.go.dev/sync#RWMutex), either N RLocks  _or_ 1 Lock can be held at any given time. I think the tests currently miss this case because they don't go through the `listen` flow.

If this is the case, one fix would be to remove the lock calls from this function and instead have `listen()` obtain a `Lock` instead of an `RLock`.

SidneyNemzer replied:
`listen` and `handleEvent` are using different mutexes, defined on the `Client` and `channel` structs respectively. Because they guard different things, there shouldn't be a deadlock here.

dkarella replied:
Ah yep that totally makes sense. It tripped me up that the calls look the same but refer to different `c`s 😁 . Thanks for clarifying!

dkarella commented on presence_channel.go:
```diff
@@ -0,0 +1,249 @@
+			c.mutex.Unlock()
+		}
+
+		select {
```
I think there's a pitfall with this `select/default` approach where it's possible for clients to miss certain events.

E.g., let's say the client listens to memberAdded on a single goroutine that takes 500ms to process a member: if member A joins at t=0ms and member B joins at t=200ms, then the message for member B joining will be dropped.


I think the reason for the `select/default` approach was to avoid blocking if there are _no_ goroutines processing the messages, right? If that's the case, I think we may need to build that into the API. I.e. we need to be able to tell _when_ a user has binded a goroutine, and only send messages then. We would also need a way for the client to unbind.

SidneyNemzer replied:
> I think the reason for the select/default approach was to avoid blocking if there are no goroutines processing the messages, right?

Yup, exactly!

> If that's the case, I think we may need to build that into the API. I.e. we need to be able to tell when a user has binded a goroutine, and only send messages then. We would also need a way for the client to unbind.

Now I'm not totally sure about this... how could we avoid the situation you mentioned above -- the consumer receives a message but doesn't listen again before the next message?

Igor noted that the current implementation wouldn't work if you try listening to `BindMember*` across multiple goroutines (since messages would only be delivered to one goroutine -- whoops). I'm going to refactor BindMemberAdded/Removed to work like `Bind` where a channel is created on each invocation.

dkarella replied:
Correct. `select/default` will only send the message if there's currently a goroutine waiting to receive the message.

E.g. in the following example:

```go
for x := range myChannel {
  // do task that takes 500ms
}
```

While that goroutine is doing the task that takes 500ms, `select/default` style sends will be dropped. Another one will only be sent successfully once its done with the loop body.

Given the same example, if the producer uses a normal send, the producer will block until the receiver is ready to receive a new message.

SidneyNemzer replied:
I guess one way is using a goroutine as a buffer like this:

```
for _, c := range chansMap {
  go func() {
    c <- message
  }()
}
```

Which is how `Bind` works

dkarella replied:
As far as supporting multiple goroutines goes, yep creating a new channel for each `Bind*` call seems like the way to go. However if we move toward blocking sends, we should also add a way to remove those channels because otherwise it's possible to deadlock if one of those consumer goroutines stops processing messages.

dkarella replied:
lmk if you want to sync up on this in a meet!

dkarella replied:
yep the buffer goroutine approach would prevent blocking the `handleEvent` goroutine. However, I think we'd still need a way to unbind the consumers and cancel those buffer goroutines, otherwise we'll have a goroutine leak instead of a deadlock – which I guess is better, but still not great 😁

SidneyNemzer replied:
Implemented a solution that uses context to unblock goroutines when `Unbind`/`UnbindMemberAdded/Removed` is used for a channel.

dkarella commented:
Note: I requested changes to prevent merging while we figure out if changes actually need to be made or not. If it turns out my concerns are wrong, then I don't actually expect there to be changes!

SidneyNemzer commented:
Force-pushed to squash changes from the reviews above

dkarella reviewed (approved):
lgtm!

dkarella commented on presence_channel.go:
```diff
@@ -0,0 +1,332 @@
+type PresenceChannel interface {
+	Channel
+
+	// BindMemberAdded returns a channel that receives a Member value when a user
```
nit: maybe we should add a note to the `Bind*` method comments saying that the corresponding `Unbind*` method must be called when done reading from the channel so the requirement is explicit

SidneyNemzer committed:
update comments

anatolym commented on presence_channel.go:
```diff
@@ -0,0 +1,334 @@
+var (
+	// ErrNotSubscribed is returned by functions that require the channel to be
+	// subscribed before being called.
+	ErrNotSubscribed = fmt.Errorf("not subscribed")
+	// ErrMissingMe indicates a presence channel was subscribed to, but the pusher
+	// server violated the message protocol by not providing a member for the
+	// current user.
+	ErrMissingMe = fmt.Errorf("missing member for current user")
```
nit:
```suggestion
	ErrNotSubscribed = errors.New("not subscribed")
	// ErrMissingMe indicates a presence channel was subscribed to, but the pusher
	// server violated the message protocol by not providing a member for the
	// current user.
	ErrMissingMe = errors.New("missing member for current user")
```

anatolym replied:
nit:
```suggestion
	ErrNotSubscribed = errors.New("not subscribed")
	// ErrMissingMe indicates a presence channel was subscribed to, but the pusher
	// server violated the message protocol by not providing a member for the
	// current user.
	ErrMissingMe = errors.New("missing member for current user")
```

SidneyNemzer committed:
switch to errors.New

SidneyNemzer committed:
add comments

SidneyNemzer committed:
avoid direct access to embedded struct

This change decouples presenceChannel from the other channel types by

- avoiding direct access to the embedded channels. A new mutex guards
  presenceChannel-specific fields, and presenceChannel methds call
  methods on the embedded structs instead of directly accessing fields.
- using a unique name on presenceChannel receivers to disambiguate from
  other channel types

Also, rewrote presenceChannel.handleEvent to use a switch to improve
readability.

anatolym commented on presence_channel.go:
```diff
@@ -0,0 +1,330 @@
+	memberRemovedChans map[chan string]chanContext
+	members            map[string]Member
+
+	membersMutex sync.RWMutex
```
nit: usually the convention is to place mutex in front of fields it guards.
```suggestion
	membersMutex       sync.RWMutex
	memberAddedChans   map[chan Member]chanContext
	memberRemovedChans map[chan string]chanContext
	members            map[string]Member
```

SidneyNemzer committed:
move mutex field

by convention, mutex should be placed before the fields it guards

Co-authored-by: Anatoly Milkov <anatolym@users.noreply.github.com>

SidneyNemzer committed:
replace context with doneChan

context wasn't really needed in this case -- it was really a wrapper for
a single channel that is closed to signal cleanup, other features of
context aren't needed. It was also a little awkward to pass around the
{context,cancel}.
